rts/motoko-rts/src/custom.rs

```diff
@@ -0,0 +1,326 @@
+        Ok(u64_from_nat64(value))
+    }
+}
+impl IntoValue for u64 {
+    unsafe fn into_value(self, _mem: &mut impl Memory) -> MotokoResult<Value> {
+        Ok(nat64_from_u64(self))
+    }
+}
+
+impl FromValue for i8 {
+    unsafe fn from_value(value: Value, _mem: &mut impl Memory) -> MotokoResult<Self> {
+        Ok(i8_from_int8(value))
+    }
+}
+impl IntoValue for i8 {
+    unsafe fn into_value(self, _mem: &mut impl Memory) -> MotokoResult<Value> {
+        Ok(int8_from_i8(self))
+    }
+}
+impl FromValue for i16 {
+    unsafe fn from_value(value: Value, _mem: &mut impl Memory) -> MotokoResult<Self> {
+        Ok(i16_from_int16(value))
+    }
+}
+impl IntoValue for i16 {
+    unsafe fn into_value(self, _mem: &mut impl Memory) -> MotokoResult<Value> {
+        Ok(int16_from_i16(self))
+    }
+}
+impl FromValue for i32 {
+    unsafe fn from_value(value: Value, _mem: &mut impl Memory) -> MotokoResult<Self> {
+        Ok(i32_from_int32(value))
+    }
+}
+impl IntoValue for i32 {
+    unsafe fn into_value(self, _mem: &mut impl Memory) -> MotokoResult<Value> {
+        Ok(int32_from_i32(self))
+    }
+}
+impl FromValue for i64 {
+    unsafe fn from_value(value: Value, _mem: &mut impl Memory) -> MotokoResult<Self> {
+        Ok(i64_from_int64(value))
+    }
+}
+impl IntoValue for i64 {
+    unsafe fn into_value(self, _mem: &mut impl Memory) -> MotokoResult<Value> {
+        Ok(int64_from_i64(self))
+    }
+}
+
+impl FromValue for bool {
+    unsafe fn from_value(value: Value, _mem: &mut impl Memory) -> MotokoResult<Self> {
+        // Do we want a function in `compile.ml` here?
+        let raw_value = value.get_raw();
+        assert!(
+            raw_value == 0 || raw_value == 1,
+            "Unexpected boolean value: {:b}",
+            raw_value
+        );
+        Ok(raw_value != 0)
+    }
+}
+impl IntoValue for bool {
+    unsafe fn into_value(self, _mem: &mut impl Memory) -> MotokoResult<Value> {
+        Ok(Value::from_raw(self as u32))
+    }
+}
+
+impl FromValue for BlobVec {
+    unsafe fn from_value(value: Value, _mem: &mut impl Memory) -> MotokoResult<Self> {
+        match value.tag() {
+            TAG_BLOB => {
+                let blob = value.as_blob();
+                let len = blob.len().as_u32();
+                Ok(BlobVec((0..len).into_iter().map(|i| blob.get(i)).collect()))
+            }
+            tag => Err(MotokoError::UnexpectedTag(tag)),
+        }
+    }
+}
+impl IntoValue for BlobVec {
+    unsafe fn into_value(self, mem: &mut impl Memory) -> MotokoResult<Value> {
+        let vec = self.0;
+        let value = alloc_blob(mem, Bytes(vec.len() as u32));
+        let blob = value.as_blob_mut();
+        let mut dest = blob.payload_addr();
+        for item in vec.into_iter() {
+            *dest = item;
+            dest = dest.add(1);
+        }
+        Ok(allocation_barrier(value))
+        // ^^ Is this allocation barrier necessary?
```

Comment:
Yes, the barrier would be essential, as otherwise a returned `Blob` could be discarded by the GC (if allocated concurrently during a GC run), potentially resulting in dangling pointers.